ethers-solc/src/resolver.rs

```diff
@@ -0,0 +1,518 @@
+    nodes: Vec<Node>,
+    /// The indices of `edges` correspond to the `nodes`. That is, `edges[0]`
+    /// is the set of outgoing edges for `nodes[0]`.
+    edges: Vec<Vec<usize>>,
+    /// index maps for a solidity file to an index, for fast lookup.
+    indices: HashMap<PathBuf, usize>,
+    /// with how many input files we started with, corresponds to `let input_files =
+    /// nodes[..num_input_files]`.
+    num_input_files: usize,
+    /// the root of the project this graph represents
+    #[allow(unused)]
+    root: PathBuf,
+}
+
+impl Graph {
+    /// Returns a list of nodes the given node index points to for the given kind.
+    pub fn imported_nodes(&self, from: usize) -> &[usize] {
+        &self.edges[from]
+    }
+
+    /// Returns all the resolved files and their index in the graph
+    pub fn files(&self) -> &HashMap<PathBuf, usize> {
+        &self.indices
+    }
+
+    /// Gets a node by index.
+    pub fn node(&self, index: usize) -> &Node {
+        &self.nodes[index]
+    }
+
+    /// Returns all files together with their paths
+    pub fn into_sources(self) -> Sources {
+        self.nodes.into_iter().map(|node| (node.path, node.source)).collect()
+    }
+
+    /// Returns an iterator that yields only those nodes that represent input files.
+    /// See `Self::resolve_sources`
+    /// This won't yield any resolved library nodes
+    pub fn input_nodes(&self) -> impl Iterator<Item = &Node> {
+        self.nodes.iter().take(self.num_input_files)
+    }
+
+    /// Resolves a number of sources within the given config
+    pub fn resolve_sources(paths: &ProjectPathsConfig, sources: Sources) -> Result<Graph> {
+        // we start off by reading all input files, which includes all solidity files from the
+        // source and test folder
+        let mut unresolved: VecDeque<(PathBuf, Node)> = sources
+            .into_par_iter()
+            .map(|(path, source)| {
+                let data = parse_data(source.as_ref());
+                (path.clone(), Node { path, source, data })
+            })
+            .collect();
+
+        // identifiers of all resolved files
+        let mut index: HashMap<_, _> =
+            unresolved.iter().enumerate().map(|(idx, (p, _))| (p.clone(), idx)).collect();
+
+        let num_input_files = unresolved.len();
+
+        // contains the files and their dependencies
+        let mut nodes = Vec::with_capacity(unresolved.len());
+        let mut edges = Vec::with_capacity(unresolved.len());
+        // now we need to resolve all imports for the source file and those imported from other
+        // locations
+        while let Some((path, node)) = unresolved.pop_front() {
+            let mut resolved_imports = Vec::with_capacity(node.data.imports.len());
+
+            // parent directory of the current file
+            let node_dir = path.parent();
+            if node_dir.is_none() {
+                continue
+            }
+            let node_dir = node_dir.unwrap();
```

Comment:
maybe more idiomatic 
```suggestion
            let node_dir = match path.parent() { Some(inner) => inner, None => continue };
```